x-pack/plugins/observability/server/routes/copilot/route.ts

```diff
@@ -40,12 +42,91 @@ const promptRoutes: {
           throw Boom.notImplemented();
         }
 
-        return client.chatCompletion.create(prompt.messages(resources.params.body as any));
+        try {
+          return await client.chatCompletion.create(prompt.messages(resources.params.body as any));
+        } catch (error: any) {
+          if (axios.isAxiosError(error) && error.response?.status === 401) {
+            throw Boom.forbidden(error.response?.statusText);
+          }
+          throw error;
+        }
       },
     });
   })
 );
 
+const trackRoute = createObservabilityServerRoute({
+  endpoint: 'POST /internal/observability/copilot/prompts/{promptId}/track',
+  params: t.type({
+    path: t.type({
+      promptId: t.string,
+    }),
+    body: t.intersection([
+      t.type({
+        responseTime: t.number,
+        messages: t.array(
+          t.intersection([
+            t.type({
+              role: t.union([
+                t.literal(ChatCompletionRequestMessageRoleEnum.System),
+                t.literal(ChatCompletionRequestMessageRoleEnum.User),
+                t.literal(ChatCompletionRequestMessageRoleEnum.Assistant),
+              ]),
+              content: t.string,
+            }),
+            t.partial({
+              name: t.string,
+            }),
+          ])
+        ),
+        response: t.string,
+      }),
+      t.partial({
+        feedbackAction: t.union([t.literal('thumbsup'), t.literal('thumbsdown')]),
+      }),
+    ]),
+  }),
+  options: {
+    tags: [],
+  },
+  handler: async (resources): Promise<void> => {
+    const {
+      params,
+      config,
+      dependencies: {
+        pluginsSetup: { core },
+      },
+    } = resources;
+
+    if (!config.aiAssistant?.enabled) {
+      throw Boom.notImplemented;
+    }
+
+    const info = await (
+      await core.getStartServices()
+    )[0].elasticsearch.client.asInternalUser.info();
+
+    const feedbackBody = {
+      cluster_id: info.cluster_uuid,
+      cluster_name: info.cluster_name,
+      prompt_name: params.path.promptId,
+      feedback_action: params.body.feedbackAction,
+      model:
+        'openAI' in config.aiAssistant.provider
+          ? config.aiAssistant.provider.openAI.model
+          : config.aiAssistant.provider.azureOpenAI.resourceName,
+      response_time: params.body.responseTime,
+      conversation: [
+        ...params.body.messages.map(({ role, content }) => ({ role, content })),
+        { role: 'system', content: params.body.response },
+      ],
+    };
+
+    await axios.post(config.aiAssistant.feedback.url, feedbackBody);
```

Comment:
Question: If this call throws an error, is it caught by core? Does it throw a generic error?

Reply:
yes, it throws an Error, which is specific to axios and handled by our own middleware.

Reply:
++ thanks!